### PR TITLE
Refactored kubemark code into provider-specific and provider-independent parts [Part-3]

### LINKED ISSUE #38967: Feature request for refactoring kubemark start-up script into provider-specific and kubernetes-specific components
Kubemark, the testing framework for kubernetes, currently has some spaghetti code specific to gce inside the cluster startup script (start-kubemark.sh).
We would ideally want to have the provider-specific code inside the script independent of the cluster kubernetes logic. The same also applies for config,
where we would want different config files for both the components (also with possibly some config that is part of the interface).

If we indeed separate these components, how should the interface between them look like? I can think of the following different levels of the interface complexity:
- One where we require the provider-specific component to define methods for fine-grained operations like creation of master-pd, master-https-server, master-ip, master-instance, etc
- One where we worry less about how (or if) we perform these fine-grained operations and rather just require the provider to start the master (with all required components that lets
  it talk to the hollow-nodes).

Any ideas on which approach is better suited (probably based on experience from kube-up)? The second approach gives more freedom to the provider about how it would
configure its components while also (possibly) giving way to more bugs / parts of the kubernetes component failing due to dependencies (which we should eliminate, if any).

@kubernetes/sig-scalability-misc @wojtek-t @gmarek

### PR DESCRIPTION
**Pull request by shyamjvs**
Fixes #38967
Applying final part of the changes in PR #39033 (which refactored kubemark code completely). The changes included in this PR are:

- Removed `test/kubemark/common.sh` and moved relevant parts of its code to the right places in start-kubemark/stop-kubemark scripts.
- Added DOCKER_REGISTRY, PROJECT, KUBEMARK_IMAGE_MAKE_TARGET variables to `/test/kubemark/cloud-provider-config.sh` to make the kubemark image push location variable wrt provider.
- Removed get-real-pod-for-hollow-node.sh as it doesn't seem to do anything useful.

@kubernetes/sig-scalability-misc @wojtek-t @gmarek

### CODE REVIEW

**Comment by shyamjvs** [comment]
cc @rthallisey

File: test/kubemark/run-e2e-tests.sh [hunk 1]
```diff
@@ -40,6 +38,6 @@ else
 	ARGS=$@
 fi
 
-go run ./hack/e2e.go -v --check_version_skew=false --test --test_args="--e2e-verify-service-account=false --dump-logs-on-failure=false ${ARGS}"
+#go run ./hack/e2e.go -v --check_version_skew=false --test --test_args="--e2e-verify-service-account=false --dump-logs-on-failure=false ${ARGS}"
```
Comment (gmarek) [inline test/kubemark/run-e2e-tests.sh]: revert this change.

File: test/kubemark/start-kubemark.sh [hunk 2]
```diff
@@ -338,8 +353,11 @@ function wait-for-hollow-nodes-to-run-or-timeout {
 }
 
 ############################### Main Function ########################################
+detect-project &> /dev/null
```
Comment (rthallisey) [inline test/kubemark/start-kubemark.sh]: Could add this as a skeleton function.  An external provider would use ```detect-project``` to search for an existing Kubernetes master, but that can also be handled in another patch.

File: test/kubemark/start-kubemark.sh [hunk 3]
```diff
@@ -19,10 +19,27 @@
 TMP_ROOT="$(dirname "${BASH_SOURCE}")/../.."
 KUBE_ROOT=$(readlink -e ${TMP_ROOT} 2> /dev/null || perl -MCwd -e 'print Cwd::abs_path shift' ${TMP_ROOT})
 
-source "${KUBE_ROOT}/test/kubemark/common.sh"
 source "${KUBE_ROOT}/test/kubemark/skeleton/util.sh"
 source "${KUBE_ROOT}/test/kubemark/cloud-provider-config.sh"
 source "${KUBE_ROOT}/test/kubemark/${CLOUD_PROVIDER}/util.sh"
+source "${KUBE_ROOT}/cluster/kubemark/${CLOUD_PROVIDER}/config-default.sh"
+source "${KUBE_ROOT}/cluster/kubemark/util.sh"
```
Comment (rthallisey) [inline test/kubemark/start-kubemark.sh]: Both ```run-e2e-tests.sh``` and ```start-kubemark.sh```  source multiple scripts that get sourced again by ```${KUBE_ROOT}/cluster/kubemark/util.sh```.  We can use ```${KUBE_ROOT}/cluster/kubemark/util.sh``` to source the common scripts between ```run-e2e-tests.sh``` and ```start-kubemark.sh```.

File: test/kubemark/resources/hollow-node_template.json [hunk 4]
```diff
@@ -106,7 +106,7 @@
 				},
 				{
 					"name": "hollow-proxy",
-					"image": "gcr.io/{{project}}/kubemark:latest",
+					"image": "{{registry}}/{{project}}/kubemark:latest",
 					"env": [
 						{
 							"name": "CONTENT_TYPE",
```
Comment (rthallisey) [inline test/kubemark/resources/hollow-node_template.json]: Farther down, the container```gcr.io/google_containers/node-problem-detector``` should also be using {{registry}}/{{project}}

File: test/kubemark/cloud-provider-config.sh [hunk 5]
```diff
@@ -15,3 +15,6 @@
 # limitations under the License.
 
 CLOUD_PROVIDER="${CLOUD_PROVIDER:-gce}"
+CONTAINER_REGISTRY="${CONTAINER_REGISTRY:-gcr.io}"
+PROJECT="${PROJECT:-}"
```
Comment (rthallisey) [inline test/kubemark/cloud-provider-config.sh]: I think we should have a default value for ```PROJECT```. If ```PROJECT``` gets passed into make as ```''```, then the image will be built as ```gcr.io//kubemark```.

File: test/kubemark/start-kubemark.sh (hunk 2 above)
Comment (shyamjvs) [inline test/kubemark/start-kubemark.sh]: I'm not really convinced that we should have this function as part of kubemark skeleton, because this function is something that should be defined upstream, which is in the cloud-provider's util. Like how we currently have it defined in `cluster/$CLOUD_PROVIDER/util.sh` (for gce, aws and gke).

File: test/kubemark/resources/hollow-node_template.json (hunk 4 above)
Comment (shyamjvs) [inline test/kubemark/resources/hollow-node_template.json]: I'm not sure about this because node-problem-detector, just like other components (apiserver, scheduler, etc) is part of proper versioned releases. Unlike kubemark which we just build and push locally to the project on-the-go (as we want it to be built against the HEAD).

Comment (shyamjvs) [inline test/kubemark/resources/hollow-node_template.json]: However, if we want to experiment with different images, we can push them to {{registry}}/{{project}} and use them within the project. So probably this is something we'd want to do in future.

File: test/kubemark/cloud-provider-config.sh (hunk 5 above)
Comment (shyamjvs) [inline test/kubemark/cloud-provider-config.sh]: We can set the project if we want to, but it's fine even otherwise as we fallback to `detect-project` setting $PROJECT.

File: test/kubemark/start-kubemark.sh (hunk 3 above)
Comment (shyamjvs) [inline test/kubemark/start-kubemark.sh]: Well believe me, the order of sourcing scripts is a very sensitive issue and after investing a few hours on this, I finally got it working. :)
It's a long answer to your question, but here's a crisp reply:
- We need to source `cluster/kubemark/${CLOUD_PROVIDER}/config-default.sh` **after** `cluster/kubemark/util.sh`, as `cluster/kubemark/util.sh` calls gce's config-default.sh somewhere inside which sets MASTER_NAME to "kubernetes-master", while we want it to be set as "kubernetes-kubemark-master"
- We need to also source `cluster/kubemark/${CLOUD_PROVIDER}/config-default.sh` **before** `cluster/kubemark/util.sh` otherwise some variables would be set by `cluster/kubemark/util.sh` that **will not be** over-written by `cluster/kubemark/${CLOUD_PROVIDER}/config-default.sh` due to assignments of the form VAR="${VAR:-"foobar"}".
- And regarding the common scripts that you've mentioned, the only common script called (at the top-level) by run-e2e-tests.sh and start-kubemark.sh is `cluster/kubemark/util.sh`, which should be fine. On the other hand, scripts `test/kubemark/skeleton/util.sh` and `test/kubemark/${CLOUD_PROVIDER}/util.sh` (which overwrites the skeleton/util.sh) should only be called in start- and stop-kubemark scripts as those are the only places where we need those functions.

Hope that answers some questions about why I did stuff the way it is done :)

Comment (shyamjvs) [inline test/kubemark/start-kubemark.sh]: We probably should make all this sourcing business cleaner, but that's for another day!

File: test/kubemark/run-e2e-tests.sh (hunk 1 above)
Comment (shyamjvs) [inline test/kubemark/run-e2e-tests.sh]: Fixed it. It's embarrassing how I unfailingly include this change always.
Maybe it's about time we fix this script to run the first command in case of being run inside a docker container and the second one otherwise. Would help save effort on local runs.
Do you think this would make sense?

File: test/kubemark/start-kubemark.sh (hunk 3 above)
Comment (rthallisey) [inline test/kubemark/start-kubemark.sh]: Thanks for the detailed answer.  I suppose we can clean it another day :).

**Comment by shyamjvs** [comment]
@gmarek @rthallisey Fixed the comments. PTAL now.

**Comment by shyamjvs** [comment]
@k8s-bot cross build this
@k8s-bot unit test this
@k8s-bot cvm gke e2e test this

**Comment by shyamjvs** [comment]
@k8s-bot unit test this
@k8s-bot cross build this
